0000-always-run-model-hook.md

```diff
@@ -0,0 +1,93 @@
+- Start Date: 2017-12-07
+- RFC PR: (leave this empty)
+- Ember Issue: (leave this empty)
+
+# Summary
+
+I'd like to propose this change: The model hook will always run, even for dynamic segments that already have their models loaded.
+
+# Motivation
+
+Currently the router will not run the model hook when transition to a route with a dynamic segment that already has its model loaded. This behavior is well intended, if we are using `{{link-to "post.show" post}}` then there's no reason to run the `post.show` model hook, the model has already been loaded!
+
+The guides do a great job [explaining](https://guides.emberjs.com/v2.17.0/routing/specifying-a-routes-model/#toc_dynamic-models) this feature.
+
+Before I started using JSON:API I thought this feature made a lot of sense, it would only fetch models when the application didn't have them. However, over the last year there's been a pattern I've seen in a couple Ember applications that this behavior complicates.
+
+Let's say we have a `posts.index` route that lists all of the posts. It's model hook:
+
+```js
+// app/routes/posts/index.js
+
+export default Ember.Route.extend({
+  model() {
+    return this.store.findAll('post');
+  }
+});
+```
+
+And then we also have a `posts.show` route that shows a specific post. When viewing a post we want to side load some additional data, like the post's comments.
+
+```js
+// app/routes/posts/show.js
+
+export default Ember.Route.extend({
+  model(params) {
+    return this.store.findRecord('post', params.post_id, {
+      include: ['comments.author'],
+      reload: true
+    });
+  }
+});
+```
+
+This creates an issue: Since the post model is already loaded, when we click a link on the index page to view a post, the model hook never runs. This causes the post to either render without comments, or worse, render with an async call to comments and then N+1 async calls to those comment's authors.
+
+To code around this behavior a popular pattern I've seen is to link to the post ID. Since the ID is not the model, Ember will run the model hook and the correct data will be loaded.  
```

Comment:
A better solution (if you really take the time to understand the problem) is to fetch the model in the model hook and fetch additional data in the afterModel hook.

I think passing the id is usually *not* the best solution.

Reply:
Or use `async model() {}` and fetch the additional data in the model hook, especially if you need to use that data to load the primary model.

Reply:
Also your model could use the links attribute as I do ofen. So all comments or related data will fetch single request.
```
    links: 
        comments:  /posts/:id/comments
```
using hook every time every transition put your app slow.

Reply:
Passing the state which is serialized in the URL and use client-side caching in the model hook (e.g. through Ember Data) is the correct solution in my opinion. The code path if entering the route directly or through a transition within the app should be the same to reduce complexity, testing cases and possible bugs.

Reply:
A second example is `transitionTo`.  A child route back transitionTo back the parent route will not fire the model hook.  This requires one to include all the data necessary in the `options`, say, for the setupController and any components in the tree to not hard fail b/c the data isn't there.